### DISCUSSION TITLE
Staubli CS8 Controller Not Reading .PGX files via USB or FTP (File Exists but not read by the Controller)

### DISCUSSION THREAD

**Discussion by PhantomStaubli** [visose/Robots#236] [Q&A] [answered]
Hello there - I am trying to upload .pgx to staubli controller (CS*) via USB and FTP. I can see the files on my PC end (e.g. FTP can see) but am unable to read it in the controller.

Can one share an example please? My controller does not read .pgx format when I save them and the save component of the plugin does not work either, so had to create my own component. 

For Staubli robots, do I need to create an entire folder structure similar to manual?  and if so, how does the robots plugin help generate it? 

I appreciate if you could help by example. Thanks

**Comment by PhantomStaubli** [comment]
I remove the toolhead from the Val3 program because I was getting the following error on the MCP: 
23:32 (File GH.dtx-line 6) ToolHead? : invalid name


Then I started getting this error in the .dtx file when running the below program: 
00:00 (File GH_000 pgx-Line 2) setPayload’ : undefined function

it appears that using the setPayload function on the DefaultTool toolhead needed to be removed since the toolhead from the program, the setPayload function will no longer be available. So I removed the setPayload function call manually for now, but I would like to reintroduce the toolhead in my programs hence why I am using a robot arm to serve an application. Below is the current .dtx file than is running ok without tool head payload. Can any one tell what is wrong with the code or GH component that I am using?

<?xml version="1.0" encoding="utf-8" ?>
<Programs xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.staubli.com/robotics/VAL3/Program/2">
  <Program name="GH_000">
    <Locals>
    </Locals>
    <Code><![CDATA[begin 
movej(joints[0], DefaultTool, mdesc0000)
movel(points[0], DefaultTool, mdesc0000)
movel(points[1], DefaultTool, mdesc0000)
movel(points[2], DefaultTool, mdesc0000)
movel(points[3], DefaultTool, mdesc0000)
movel(points[4], DefaultTool, mdesc0000)
movel(points[5], DefaultTool, mdesc0000)
movel(points[6], DefaultTool, mdesc0000)
movel(points[7], DefaultTool, mdesc0000)
end]]></Code>
  </Program>
</Programs>

**Reply by visose** [reply]
Not sure if I follow. If you get an error saying `invalid name` you need to change the name of the tool to a valid one in the CreateTool component. For example, remove spaces and characters that are not letters. Keep the name short.

**Comment by PhantomStaubli** [comment]
It says invalid name because the value of the setPayload is not assigned. In this case we need to assign four parameters to the program: tool name, tool weight, tool cog, inertia. That is the better explanation of the error I am getting. Please see the attachment. If I manually remove the line from the program, it runs on the robot, but the payload parameters is an important one. It may perhaps be a bug in the Val3 wrapper code for the "Create Program" component or the DLL file? Please shed some lights. Thanks! 
![Screenshot 2023-06-25 at 8 04 31 AM](https://github.com/visose/Robots/assets/137115823/1c43b617-79f4-433c-a0be-c36f758befdf)

**Reply by PhantomStaubli** [reply]
and just to be clear, I have already passed down weight and tool name via the Create Tool Component and Tool Parameter, and I see them in the .dtx file generated. For some reason the controller does not read the four inputs and gives the first error I shared unless I remove the line.

**Comment by visose** [comment]
Could you attach a minimal reproducible example?

**Reply by PhantomStaubli** [reply]
https://www.dropbox.com/scl/fo/fke783jusd3m964n8yx90/h?dl=0&rlkey=u4cdm22iyrb0ct5yvzr56syai

**Reply by visose** [reply]
A few issues:
- It looks like you're using a custom robot library not available online, please include it aswell in the zip file.
- The Grasshopper file has references to Rhino objects that don't exist in the attached Rhino file. Make sure there are no broken references.
- The 20 MB Rhino document appears to be empty, the size is due to unused block definitions. Please purge the file before attaching.
- Use meaningful names such as file and layer names to help troubleshoot the issue.

**Comment by PhantomStaubli** [comment]
Thanks, but I think I know what the problem is. I just need to know how to pass down the following to the setPayload and add being and end before and after it. right know the program only generates one line which is "setPayload(flange, l_nMass, l_trGrCenter, l_nMatrix)". Can you also let me know how I could edit the robot program component myself? 

begin
  l_nMass = 1.0000
  l_trGrCenter = {75.0000 , 75.0000 , 75.0000, 0, 0, 0}
  l_nMatrix[0,0]=0.0038
  l_nMatrix[0,1]=0.0000
  l_nMatrix[0,2]=0.0000
  l_nMatrix[1,0]=0.0000
  l_nMatrix[1,1]=0.0038
  l_nMatrix[1,2]=0.0000
  l_nMatrix[2,0]=0.0000
  l_nMatrix[2,1]=0.0000
  l_nMatrix[2,2]=0.0038

  setPayload(flange, l_nMass, l_trGrCenter, l_nMatrix)
end

**Reply by visose** [reply]
Are you copying all files generated with the `Save Program` component to the controller? The variables used by `setPayload` are expected to be read from the generated .dtx file.

**Reply by PhantomStaubli** [reply]
Yes. I am copying all the applications generated including test_000.pgx, test.pjx, test.dtx, start.pgx, end.pgx. but the test_000/pgx is not reading the input data fro the .dtx file. I thought the format written in my previous post would be the solution but I still agree with your comment. 

Below are my settings in the .dtx file.

<Data name="Inertia" access="private" xsi:type="array" type="num" size="1">
<Value key="0" value="0"/>
</Data>
<Data name=“Hand” access="private" xsi:type="array" type="tool" size="1">
<Value key="0" x="-0.001" y="-0.005" z="122.227" rx="0" ry="0" rz="0" fatherId="flange[0]"/>
</Data>
<Data name="Hand_C" access="private" xsi:type="array" type="trsf" size="1">
<Value key="0" x="-0.001" y="-0.005" z="122.227"/>
</Data>
<Data name="Hand_W" access="private" xsi:type="array" type="num" size="1">
<Value key="0" value="0.2"/>


The .pgx file starts like the following but apparently not reading the .dtx data written above:

<?xml version="1.0" encoding="utf-8" ?>
<Programs xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.staubli.com/robotics/VAL3/Program/2">
<Program name="GH_000">
<Locals>
</Locals>
<Code><![CDATA[begin
setPayload(Hand, Hand_W, Hand_C, Inertia)



Any advise would be highly appreciated. Thanks again!

**Reply by PhantomStaubli** [reply]
![Screenshot 2023-07-07 at 8 09 52 PM](https://github.com/visose/Robots/assets/137115823/5df6a9a8-8335-4eb9-b874-e6bfcc50856f)

**Comment by PhantomStaubli** [comment]
Inertia need to be an array 3,3 of num like this: 
<Data name="nInertia" access="private" xsi:type="array" type="num" size="3 3" />

is there any way to pass this down to the component? If, so how and which input component? I could not find it in the documentation or components in GH. Thanks again.

**Comment by PhantomStaubli** [comment]
@visose I did not hear from you since I wrote last time. Hope all is well. The following error message on the Staubli MCP is still very persistent:

​(File Test_000.pgx-Line 2) ‘setPayload' undefined function

Below are the status of my application architecture: 

1) the 'setPayload' function is correctly defined and spelled in the '.dtx' file.​ 
2) the '.dtx' file is properly included and linked within the application and is accessible to the '.pgx' file.

the folder structure and file locations are correct and the MCP can access the coordination points from .dtx folder but unable to read the following function from it:

<?xml version="1.0" encoding="utf-8" ?>
<Database xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.staubli.com/robotics/VAL3/Data/2">
  <Datas>
    <Data name="Inertia" access="private" xsi:type="array" type="num" size="1">
        <Value key="0" value="0"/>
    </Data>
    <Data name="Toolhead” access="private" xsi:type="array" type="tool" size="1">
        <Value key="0" x="-0.001" y="-0.005" z="122.227" rx="0" ry="0" rz="0" fatherId="flange[0]"/>
    </Data>
    <Data name="Toolhead_C" access="private" xsi:type="array" type="trsf" size="1">
        <Value key="0" x="-0.001" y="-0.005" z="122.227"/>
    </Data>
    <Data name="Toolhead_W" access="private" xsi:type="array" type="num" size="1">
        <Value key="0" value="0.2"/>
​
​
​This is the section of the .pgx file where the error is coming from: 

 <?xml version="1.0" encoding="utf-8" ?> <Programs xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.staubli.com/robotics/VAL3/Program/2"> <Program name="Cone_000"> <Locals> </Locals> <Code><![CDATA[begin setPayload(Toolhead​​, Toolhead​​_W, Toolhead​​_C, Inertia)​


I am running out of time and need to figure this out. Please kindly advise. Thanks very much for your time spent.

**Reply by visose** [reply]
I just made a simple test and it works, the program runs on a SC8 controller with no issues. Attached is the test file. It requires having the Bartlett robot library.
[StaubliTest.zip](https://github.com/visose/Robots/files/12136582/StaubliTest.zip)

**Comment by visose** [comment]
I just read this on your top comment:
>the save component of the plugin does not work either, so had to create my own component

This seems like a red flag. Why is the save component not working? Can't you save the files to a folder in your PC?

**Reply by PhantomStaubli** [reply]
The save program works fine. Let me test your GH and get back to you. I appreciate your time spent once again.

**Comment by PhantomStaubli** [comment]
still getting error on the MCP inside CS8 Controller. Also noticed there was not Tool defined in the example. See attached. Problem persists.

![Screenshot 2023-07-23 at 11 04 58 AM](https://github.com/visose/Robots/assets/137115823/b763342c-0e3b-4b62-8c2c-2c08aaa5d514)

**Reply by visose** [reply]
Sorry, if I'm not able to reproduce the issue there's not much else I can do.

**Comment by PhantomStaubli** [comment]
problem solved by updating val3 to new version

**Comment by PhantomStaubli** [comment]
[visose](https://github.com/visose) how do I get smooth motions by introducing declarations in GH components? Right now, the arm moves like crazy jagged and stops and accelerates again. I need to have smooth motions after reaching each point and I have hundred or perhaps thousands of points. Another thing I noticed is that the controller freezes with that many points in it. Any suggestions? Please advise. Thanks.

**Reply by visose** [reply, marked as answer]
Use the zone input in CreateTarget for smooth motions. Use the multifile input in CreateProgram to not freeze the controller. Search this forum and check the wiki for more info on these.

**Reply by PhantomStaubli** [reply]
@visose Could you please point me in the wiki to learn more about multifile and gow to add a minimum reproducible gh file integers? I tried assigning integers and the program component gives an error. Im using a staubli robot. Thanks very much for your time spent on this...

**Reply by visose** [reply]
I replied in https://github.com/visose/Robots/discussions/72#discussioncomment-8889449
